src/otx/core/model/entity/detection.py

```diff
@@ -0,0 +1,138 @@
+    compatible for OTX pipelines.
+    """
+
+    def __init__(self, config: DictConfig) -> None:
+        self.config = config
+        super().__init__()
+
+    def _create_model(self) -> nn.Module:
+        # import mmdet.models as _
+        from mmdet.registry import MODELS
+        from mmengine.registry import MODELS as MMENGINE_MODELS
+
+        # RTMDet-Tiny has bug if we pass dictionary data_preprocessor to MODELS.build
+        # We should inject DetDataPreprocessor to MMENGINE MODELS explicitly.
+        det = MODELS.get("DetDataPreprocessor")
+        MMENGINE_MODELS.register_module(module=det)
+
+        try:
+            model = MODELS.build(convert_conf_to_mmconfig_dict(self.config, to="tuple"))
+        except AssertionError:
+            model = MODELS.build(convert_conf_to_mmconfig_dict(self.config, to="list"))
+
+        return model
+
+    def _customize_inputs(self, entity: DetBatchDataEntity) -> dict[str, Any]:
+        from mmdet.structures import DetDataSample
+        from mmengine.structures import InstanceData
+
+        mmdet_inputs: dict[str, Any] = {}
+
+        mmdet_inputs["inputs"] = entity.images  # B x C x H x W PyTorch tensor
+        mmdet_inputs["data_samples"] = [
+            DetDataSample(
+                metainfo={
+                    "img_id": img_info.img_idx,
+                    "img_shape": img_info.img_shape,
+                    "ori_shape": img_info.ori_shape,
+                    "pad_shape": img_info.pad_shape,
+                    "scale_factor": img_info.scale_factor,
+                },
+                gt_instances=InstanceData(
+                    bboxes=bboxes,
+                    labels=labels,
+                ),
+            )
+            for img_info, bboxes, labels in zip(
+                entity.imgs_info,
+                entity.bboxes,
+                entity.labels,
+            )
+        ]
+        preprocessor: DetDataPreprocessor = self.model.data_preprocessor
+        # Don't know why but data_preprocessor.device is not automatically
+        # converted by the pl.Trainer's instruction unless the model parameters.
+        # Therefore, we change it here in that case.
+        if preprocessor.device != (
+            model_device := next(self.model.parameters()).device
+        ):
+            preprocessor = preprocessor.to(device=model_device)
+            self.model.data_preprocessor = preprocessor
+
+        mmdet_inputs = preprocessor(data=mmdet_inputs, training=self.training)
+
+        mmdet_inputs["mode"] = "loss" if self.training else "predict"
+
+        return mmdet_inputs
+
+    def _customize_outputs(
+        self,
+        outputs: Any,  # noqa: ANN401
+    ) -> Union[DetBatchPredEntity, OTXBatchLossEntity]:
+        from mmdet.structures import DetDataSample
+
+        if self.training:
+            if not isinstance(outputs, dict):
+                raise TypeError(outputs)
+
+            losses = OTXBatchLossEntity()
```

Comment:
Why not simplify the code by using a type hint such as `losses: Dict[str, Tensor]` instead of creating a new entity?

Reply:
See my above comment